### ISSUE TITLE
Cray compiler

### ISSUE THREAD

**Issue by jeffhammond** [elemental/Elemental#185]
As of version 8.4, Cray compilers should support C++11 (http://en.cppreference.com/w/cpp/compiler_support).

# CMake invocation

```
rm -rf * && cmake .. \
-DCMAKE_INSTALL_PREFIX=$HOME/ELEMENTAL/install-cray \
-DCMAKE_CXX_COMPILER=CC \
-DCMAKE_C_COMPILER=cc \
-DCMAKE_Fortran_COMPILER=ftn \
-DMPI_CXX_COMPILER=CC \
-DMPI_C_COMPILER=cc  \
-DMPI_Fortran_COMPILER=ftn \
-DMPI_C_INCLUDE_PATH=$CRAY_MPICH_DIR/include \
-DCMAKE_CXX_FLAGS="-h std=c++11 -h omp " \
-DCMAKE_C_FLAGS="-h std=c11 -h omp " \
-DMATH_LIBS="-libsci_cray_mpi_mp -lsci_cray_mp " \
-DEL_HYBRID=TRUE \
-DEL_DISABLE_PARMETIS=TRUE \
-DCMAKE_BUILD_TYPE=Debug \
-DEL_USE_64BIT_INTS=ON \
-DEL_USE_64BIT_BLAS_INTS=OFF \
-DINSTALL_PYTHON_PACKAGE=OFF \
-DEL_DISABLE_QUAD=TRUE \
-DCMAKE_CXX_COMPILER_ID=Cray \
-DCMAKE_C_COMPILER_ID=Cray \
-DCMAKE_Fortran_COMPILER_ID=Cray
```

# Code changes

```
diff --git a/cmake/modules/FindCXXFeatures.cmake b/cmake/modules/FindCXXFeatures.cmake
index b720345..d1cfc13 100644
--- a/cmake/modules/FindCXXFeatures.cmake
+++ b/cmake/modules/FindCXXFeatures.cmake
@@ -63,6 +63,11 @@ elseif (CMAKE_CXX_COMPILER_ID MATCHES "(Borland|Watcom)")
     # No C++11 flag for those compilers, but check_cxx_compiler_flag()
     # can't detect because they either will not always complain (Borland)
     # or will hang (Watcom).
+elseif (CMAKE_CXX_COMPILER_ID MATCHES "Cray")
+    test_set_flag("-h std=c++11" CXX11) 
+    if (NOT CXX11_COMPILER_FLAGS)
+        message(FATAL_ERROR "Could not use -h std=c++11 with Cray C/C++")
+    endif()
 elseif (CMAKE_CXX_COMPILER_ID STREQUAL "Intel" AND WIN32)
     # The Intel compiler on Windows may use these flags.
     test_set_flag("/Qstd=c++11" CXX11)
```

**Comment by jeffhammond** [comment]
I'll push CMake patches later but we are blocked by an ICE:
```
CC-2116 crayc++: INTERNAL  
  "/opt/cray/cce/8.5.1/CC/x86-64/lib/ccfe" was terminated due to receipt of signal 013:  Segmentation fault.
```

For posterity, this is how I built:
```
cmake .. \
-DCMAKE_INSTALL_PREFIX=$HOME/ELEMENTAL/install-cray \
-DBUILD_SHARED_LIBS=OFF \
-DCMAKE_CXX_COMPILER_ID=Cray \
-DCMAKE_C_COMPILER_ID=Cray \
-DCMAKE_Fortran_COMPILER_ID=Cray \
-DCMAKE_CXX_COMPILER=CC \
-DCMAKE_C_COMPILER=cc \
-DCMAKE_Fortran_COMPILER=ftn \
-DMPI_CXX_COMPILER=CC \
-DMPI_C_COMPILER=cc  \
-DMPI_Fortran_COMPILER=ftn \
    -DMPI_C_INCLUDE_PATH=$CRAY_MPICH_DIR/include \
    -DMPI_C_COMPILE_FLAGS="" \
    -DMPI_C_LINK_FLAGS="" \
    -DMPI_C_LIBRARIES="" \
-DCMAKE_CXX_FLAGS="-h std=c++11 -h omp " \
-DCMAKE_C_FLAGS="-h std=c11 -h omp " \
-DMATH_LIBS="-L${CRAY_LIBSCI_PREFIX_DIR}/lib -lsci_cray_mpi_mp -lsci_cray_mp " \
-DEL_HYBRID=TRUE \
-DEL_DISABLE_PARMETIS=TRUE \
-DCMAKE_BUILD_TYPE=Debug \
-DEL_USE_64BIT_INTS=ON \
-DEL_USE_64BIT_BLAS_INTS=OFF \
-DINSTALL_PYTHON_PACKAGE=OFF \
-DEL_DISABLE_QUAD=TRUE
```

**Comment by poulson** [comment]
Thanks for noticing this Jeff! I'll follow up in more detail on Gitter.

**Comment by rhl-** [comment]
Addressed by #186

**Comment by PierreCarrier** [comment]
Although this post is rather old, I tried the build on a current XC40 and it still worked. I would suggest to include some makefiles inside the Elemental-master/examples directory. This one that I build seems to be working:

==============================
CC       = CC
BASE_EL= ${HOME}/ELEMENTAL/install-cray
CFLAG    =  -O2 -I${BASE_EL}/include -I${BASE_EL}/build/include -I${BASE_EL}/external/suite_sparse/include

.SUFFIXES : .o .cpp
.cpp.o :
        $(CC) $(CFLAG) -c $<

MAIN = HermitianEig.o
OBJCT = \

./elemental: $(OBJCT) $(MAIN)
        $(CC) $(CFLAG) -o ./elemental $(MAIN) $(OBJCT) -L${BASE_EL}/lib64 -lEl -lpmrrr
clean:
        rm -f *.o
cleanall:
        rm -f *.o elemental
====================================
